code/modules/mob/living/basic/trader/trader_ai.dm

```diff
@@ -0,0 +1,120 @@
+/datum/ai_planning_subtree/setup_shop
+	/// What do we do in order to offer our deals?
+	var/datum/ai_behavior/setup_shop/setup_shop_behavior = /datum/ai_behavior/setup_shop
+
+/datum/ai_planning_subtree/setup_shop/SelectBehaviors(datum/ai_controller/controller, seconds_per_tick)
+	//if we already have a shop spot, return
+	if(controller.blackboard_key_exists(BB_SHOP_SPOT))
+		return
+
+	//if we don't have a costurmer to greet, look for one
+	if(!controller.blackboard_key_exists(BB_FIRST_CUSTOMER))
+		controller.queue_behavior(/datum/ai_behavior/find_and_set/conscious_person, BB_FIRST_CUSTOMER, /mob/living/carbon/human)
+		return
+
+	//we have our first customer, time to tell them about incredible deals
+	controller.queue_behavior(setup_shop_behavior, BB_FIRST_CUSTOMER)
+	return SUBTREE_RETURN_FINISH_PLANNING
+
+///The ai will create a shop the moment they see a potential costumer
+/datum/ai_behavior/setup_shop
+
+/datum/ai_behavior/setup_shop/setup(datum/ai_controller/controller, target_key)
+	. = ..()
+	var/obj/target = controller.blackboard[target_key]
+	if(QDELETED(target))
+		return FALSE
+
+/datum/ai_behavior/setup_shop/perform(seconds_per_tick, datum/ai_controller/controller, target_key)
+	. = ..()
+
+	//we lost track of our costumer, abort
+	if(!controller.blackboard_key_exists(target_key))
+		finish_action(controller, FALSE, target_key)
+		return
+
+	var/mob/living/basic/living_pawn = controller.pawn
+
+	living_pawn.say("Welcome to my shop, friend!")
+	var/shop_type_path =  controller.blackboard[BB_SHOP_SPOT_TYPE]
+	var/obj/shop_spot = new shop_type_path(living_pawn.loc)
+	shop_spot.dir = living_pawn.dir
+	controller.set_blackboard_key(BB_SHOP_SPOT, shop_spot)
+	playsound(living_pawn, controller.blackboard[BB_SHOP_SOUND], 50, TRUE)
+
+	var/turf/sign_turf
+
+	sign_turf = try_find_valid_spot(living_pawn.loc, turn(shop_spot.dir, -90))
+	if(!sign_turf) //No space to my left, lets try right
+		sign_turf = try_find_valid_spot(living_pawn.loc, turn(shop_spot.dir, 90))
+
+	if(sign_turf)
+		var/obj/sign = controller.blackboard[BB_SHOP_SIGN]
+		if(QDELETED(sign))
+			var/sign_type_path =  controller.blackboard[BB_SHOP_SIGN_TYPE]
+			var/obj/new_sign = new sign_type_path(sign_turf)
+			controller.set_blackboard_key(BB_SHOP_SIGN, new_sign)
+			do_sparks(3, FALSE, new_sign)
+		else
+			do_teleport(sign,sign_turf)
+
+	controller.clear_blackboard_key(BB_FIRST_CUSTOMER)
+
+	finish_action(controller, TRUE, target_key)
+
+///Version of setup show where the trader will run at you to assault you with incredible deals
+/datum/ai_planning_subtree/setup_shop/jumpscare
+	setup_shop_behavior = /datum/ai_behavior/setup_shop/jumpscare
+
+/datum/ai_behavior/setup_shop/jumpscare
+	behavior_flags = AI_BEHAVIOR_REQUIRE_MOVEMENT | AI_BEHAVIOR_REQUIRE_REACH
+
+/datum/ai_behavior/setup_shop/jumpscare/setup(datum/ai_controller/controller, target_key)
+	. = ..()
+	if(.)
+		set_movement_target(controller, controller.blackboard[target_key])
```

Comment:
Can you move these below the two following generic procs to keep the parent/child relationship clearer?